File: src/java/magmac/api/head/SingleHead.java
package magmac.api.head;

import magmac.api.None;
import magmac.api.Option;
import magmac.api.Some;

public class SingleHead<T> implements Head<T> {
    private final T element;
    private boolean retrieved;

    public SingleHead(T element) {
        this.element = element;
        this.retrieved = false;
    }

    @Override
    public Option<T> next() {
        if (this.retrieved) {
            return new None<>();
        }

        this.retrieved = true;
        return new Some<>(this.element);
    }
}

File: src/web/magmac/api/head/SingleHead.ts
import { None } from "../../../magmac/api/None";
import { Option } from "../../../magmac/api/Option";
import { Some } from "../../../magmac/api/Some";
export class SingleHead<T> {
	SingleHead(element : T) : public {break;break;;}
	next() : Option<T> {if(true){ return new None<>( );;}break;return new Some<>( 0.element);;}
}
